studies/tests.py

```diff
@@ -0,0 +1,246 @@
+            ),
+            targets,
+        )
+
+        # Sanity check - participant #2 isn't suddenly the parent of child #1...
+        self.assertFalse(
+            any(
+                mt.user_id == self.participant_two.id
+                and mt.child_id == self.child_one.id
+                for mt in targets
+            )
+        )
+        # Check other way around as well
+        self.assertFalse(
+            any(
+                mt.user_id == self.participant_one.id
+                and mt.child_id in (self.child_two.id, self.child_three.id)
+                for mt in targets
+            )
+        )
+
+    def test_target_creation_e2e(self):
+        targets = list(
+            _segmented_by_study(
+                _validated(_deserialized(_grouped_by_user(potential_message_targets())))
+            )
+        )
+
+        self.assertEqual(len(targets), 1)
+
+        user, study_child_mapping = targets[0]
+
+        self.assertEqual(user, self.participant_two)
+
+        self.assertDictEqual(
+            study_child_mapping,
+            {
+                self.study_one: [self.child_three],
+                self.study_two: [self.child_two, self.child_three],
+            },
+        )
+
+    def test_correct_message_structure(self):
+        target_email_structure = TARGET_EMAIL_TEMPLATE.format(
+            base_url=settings.BASE_URL, study_id=self.study_two.id
+        )
+
+        message_object: Message = Message.send_announcement_email(
+            self.participant_two, self.study_two, [self.child_two, self.child_three]
+        )
+        self.assertMultiLineEqual(message_object.body, target_email_structure)
+        self.assertEqual(
+            message_object.subject,
+            "New study available on Lookit: The Most Fake Study Ever",
+        )
+        self.assertEqual(list(message_object.recipients.all()), [self.participant_two])
+        self.assertEqual(
+            list(message_object.children_of_interest.all()),
+            [self.child_two, self.child_three],
+        )
+        self.assertEqual(message_object.related_study, self.study_two)
+        self.assertIsNone(message_object.sender)
+
+    def test_study_excluded_from_targets_after_message(self):
+        Message.send_announcement_email(
+            self.participant_two, self.study_one, [self.child_three]
+        )
+        # New targets should have only study two targets
+        targets = list(
+            _segmented_by_study(
+                _validated(_deserialized(_grouped_by_user(potential_message_targets())))
+            )
+        )
+
+        # First level is the same - it's the same user...
+        self.assertEqual(len(targets), 1)
+        user, study_child_mapping = targets[0]
+        self.assertEqual(user, self.participant_two)
+
+        # ... But no study one this time.
+        self.assertDictEqual(
+            study_child_mapping, {self.study_two: [self.child_two, self.child_three]},
+        )
```

Comment:
Would propose a few additional tests (or additions to existing tests) here:
* No targets generated for Study that is active but non-discoverable or paused but discoverable
* No targets generated for User with email_new_studies set to false
* No (user, study, child) target if child is older than max age for study or younger than min age for study
* No (user, study, child) target if child does not meet criteria expression (e.g. child has to speak Spanish to participate)

Reply:
Done. Was thinking that we should explicitly test criteria expression stuff in that suite, but I guess it makes sense to see that the checker was called properly.

Reply:
🎈 🍰 am I doing this right